lib/netplay/port_mapping_manager.cpp

```diff
@@ -62,13 +71,644 @@ void PlumMappingCallback(int mappingId, plum_state_t state, const plum_mapping_t
 	}
 }
 
-void PlumLogCallback(plum_log_level_t /*level*/, const char* message)
+void PlumLogCallback(plum_log_level_t level, const char* message)
 {
-	debug(LOG_NET, "LibPlum message: %s", message);
+	code_part wz_log_level = LOG_NET;
+	switch (level)
+	{
+	case PLUM_LOG_LEVEL_FATAL:
+	case PLUM_LOG_LEVEL_ERROR:
+	case PLUM_LOG_LEVEL_WARN:
+		wz_log_level = LOG_INFO;
+		break;
+	default:
+		break;
+	}
+	if (enabled_debug[wz_log_level])
+	{
+		std::string msg = (message) ? message : "";
+		wzAsyncExecOnMainThread([wz_log_level, msg]() {
+			_debug(__LINE__, wz_log_level, "PlumLogCallback", "%s", msg.c_str());
+		});
+	}
 }
 
 } // anonymous namespace
 
+class PortMappingImpl_LibPlum : public PortMappingImpl
+{
+public:
+	virtual ~PortMappingImpl_LibPlum();
+	virtual bool init() override;
+	virtual bool shutdown() override;
+	virtual int create_port_mapping(uint16_t port, PortMappingInternetProtocol protocol) override;
+	virtual bool destroy_port_mapping(int mappingId) override;
+	virtual PortMappingImpl::Type get_impl_type() const override;
+	virtual int get_discovery_timeout_seconds() const override;
+private:
+	bool m_isInit = false;
+};
+
+PortMappingImpl_LibPlum::~PortMappingImpl_LibPlum()
+{
+	try
+	{
+		shutdown();
+	}
+	catch(...) // Don't let any exceptions escape the dtor.
+	{}
+}
+
+PortMappingImpl::Type PortMappingImpl_LibPlum::get_impl_type() const
+{
+	return PortMappingImpl::Type::Libplum;
+}
+
+int PortMappingImpl_LibPlum::get_discovery_timeout_seconds() const
+{
+	return 10; // 10 seconds for Libplum, to ensure we don't wait too long before falling-back to Miniupnpc
+}
+
+bool PortMappingImpl_LibPlum::init()
+{
+	plum_config_t config;
+	memset(&config, 0, sizeof(config));
+	config.log_level = PLUM_LOG_LEVEL_INFO;
+	config.log_callback = &PlumLogCallback;
+	auto res = plum_init(&config);
+	m_isInit = (res == PLUM_ERR_SUCCESS);
+	return res == PLUM_ERR_SUCCESS;
+}
+
+bool PortMappingImpl_LibPlum::shutdown()
+{
+	if (!m_isInit)
+	{
+		return false;
+	}
+	plum_cleanup();
+	m_isInit = false;
+	return true;
+}
+
+int PortMappingImpl_LibPlum::create_port_mapping(uint16_t port, PortMappingInternetProtocol protocol)
+{
+	plum_mapping_t m;
+	memset(&m, 0, sizeof(m));
+	m.protocol = PLUM_IP_PROTOCOL_TCP;
+	m.internal_port = port;
+	m.external_port = port; // suggest an external port the same as the internal port (the router may decide otherwise)
+
+	auto mappingId = plum_create_mapping(&m, PlumMappingCallback);
+	return mappingId;
+}
+
+bool PortMappingImpl_LibPlum::destroy_port_mapping(int mappingId)
+{
+	const auto result = plum_destroy_mapping(mappingId);
+	return result == PLUM_ERR_SUCCESS;
+}
+
+// MARK: - Minupnpc implementation
+
+#if (defined(__GNUC__) || defined(__clang__)) && !defined(__INTEL_COMPILER)
```

Comment:
Let's keep all headers at the top of the file - it's always easier to deal with them this way.